File: gen_vapid.js
const { createECDH } = require('crypto');

// prime256v1 = P-256
const ecdh = createECDH('prime256v1');
ecdh.generateKeys();

// getPublicKey() returns 65 bytes: 04 || x || y (uncompressed point)
const pubBytes = ecdh.getPublicKey(); // Buffer
const pubB64 = pubBytes.toString('base64url');

// getPrivateKey() returns 32 bytes raw scalar
const privBytes = ecdh.getPrivateKey(); // Buffer
const privB64 = privBytes.toString('base64url');

console.log('PUBLIC:', pubB64);
console.log('PRIVATE:', privB64);
console.log('pub length:', pubBytes.length, '(should be 65)');
console.log('priv length:', privBytes.length, '(should be 32)');
console.log('first pub byte:', pubBytes[0].toString(16), '(should be 04)');

File: gen_vapid.py
from cryptography.hazmat.primitives.asymmetric.ec import generate_private_key, SECP256R1
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat, PrivateFormat, NoEncryption
import base64

private_key = generate_private_key(SECP256R1())

# Public key: uncompressed point (65 bytes)
pub_bytes = private_key.public_key().public_bytes(Encoding.X962, PublicFormat.UncompressedPoint)
pub_b64 = base64.urlsafe_b64encode(pub_bytes).rstrip(b'=').decode()

# Private key: DER -> extract raw 32-byte scalar (last 32 bytes of the private key number)
priv_num = private_key.private_numbers().private_value
priv_bytes = priv_num.to_bytes(32, 'big')
priv_b64 = base64.urlsafe_b64encode(priv_bytes).rstrip(b'=').decode()

print("PUBLIC:", pub_b64)
print("PRIVATE:", priv_b64)
print()
print("Len public bytes:", len(pub_bytes), "(should be 65)")
print("Len private bytes:", len(priv_bytes), "(should be 32)")
